spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+This can be made more expressive by defining it as `with: String|((RegexMatch) -> String)`.
+Although this would indeed make `pkl:Project` more expressive, the API and CLI definition cases are less amenable to this solution.
+In the API case, users would have to programmatically construct Truffle node instances.
+The CLI would require significant parsing support, or impose prohibitive string escaping requirements.
+Introducing arbitrary functions into the evaluation of modules from `pkl:Project` also has implications for performance and error messages.
+All of these are considered detrimental to the user experience.
+
+
+== Usage examples
+
+=== Context-dependent rewrite rule sets
+
+If some rewrites only apply to tests, or parts of a repository, they could all be defined on the CLI.
+This can become cumbersome to maintain, since this means they are defined in (typically non-Pkl) build configuration files.
+However, since a `PklProject` may amend another `PklProject` file, rewrite rules can be kept in Pkl modules that amend a common root `PklProject` file.
+
+Taking testing as an example, you can keep the following file structure:
+----
+.
+├── PklProject         // <1>
+├── src
+│   └── foo.pkl
+└── test
+    ├── foo_test.pkl
+    └── PklProject    // <2>
+----
+<1> Defines common URI rewrite rules and `amends "pkl:Project"`.
+<2> Defines URI rewrite rules used only for testing and `amends "..."`.
+
+Running `pkl test foo_test.pkl` applies all the rules for the testing context.
+
+=== Test rewrite rules
+
+No new facilities are required for `pkl:test` to cover rewrite rules.
+
+An easy way to perform positive tests (testing whether rewrite rules _have_ been applied) is to disallow all module keys and to test that the expected rewritten URIs occur in the exception.
+Negative tests are similar, except the assertion is that the exception contains the original URI.
+For example, if `"github.com"` is rewritten to `"github.example.com"`, and URIs pointing at `example.com` should not be rewritten, the following test passes when run with `pkl test --allowed-modules file test.pkl`:
+
+[source,pkl]
+----
+amends "pkl:test"
+
+facts {
+  ["github.com URIs are pointed at the example.com GHE instance"] {
+    module.catch(() -> import("https://github.com/foo/bar.pkl"))
+      .contains("https://github.example.com/foo/bar.pkl")
+  }
+  ["example.com URIs are *not* rewritten"] {
+    module.catch(() -> import("https://example.com/foo/bar.pkl"))
+      .contains("https://example.com/foo/bar.pkl")
+  }
+}
+----
+
+=== DENY default rule
+
+When URI rewrite rules are used to ensure no remote resources are used at all, a rule to check "all URIs not touched by any of the rewrite rules are rejected" can be expressed.
+This rule is based on the fact that it matches at zero-width, so no other rule has a lower score:
+
+[source,pkl]
+----
+  new UriRewrite {
+    replace = Regex("^")
+    with = throw("URI found that is uncovered by any of the rewrite rules")
+  }
```

Comment:
Not sure if this can work--in order to configure the evaluator at runtime, all of the rewrite rules need to be consumed (ergo, deep forced).